### PR TITLE
ship log to cloudwatch via syslog

### PR DESCRIPTION
**Pull request by BumbleFeng**
Add Syslog as new logger handler
Add [syslog-cloudwatch-bridge](https://github.com/app-sre/syslog-cloudwatch-bridge) as sidecar to ship log to cloudwatch
https://issues.redhat.com/browse/APPSRE-1664

### CODE REVIEW

File: reconcile/cli.py [hunk 1]
```diff
@@ -219,6 +220,11 @@ def init_log_level(log_level):
     logging.basicConfig(format=format,
                         datefmt='%Y-%m-%d %H:%M:%S',
                         level=level)
+    syslog = SysLogHandler(address=('localhost', 5014))
```
Comment (tparikh) [inline reconcile/cli.py]: make host & port configurable from `config.toml`

File: helm/qontract-reconcile/templates/template.yaml [hunk 2]
```diff
@@ -141,15 +141,68 @@ objects:
...
+                secretKeyRef:
+                  name: ${CLOUDWATCH_SECRET}
+                  key: log_group_name
+          resources:
+            requests:
+              memory: 30Mi
+              cpu: 15m
+            limits:
+              memory: 60Mi
+              cpu: 25m
```
Comment (tparikh) [inline helm/qontract-reconcile/templates/template.yaml]: Are you sure this is enough? How about `100m`?

File: helm/qontract-reconcile/templates/template.yaml [hunk 3]
```diff
@@ -141,15 +141,68 @@ objects:
...
+              valueFrom:
+                secretKeyRef:
+                  name: ${CLOUDWATCH_SECRET}
+                  key: log_group_name
+          resources:
+            requests:
+              memory: 30Mi
+              cpu: 15m
+            limits:
+              memory: 60Mi
```
Comment (tparikh) [inline helm/qontract-reconcile/templates/template.yaml]: Are you sure this is enough?

File: reconcile/cli.py (hunk 1 above)
Comment (tparikh) [reply reconcile/cli.py]: What happens if there is no syslog server like in dev env?

Comment (BumbleFeng) [reply reconcile/cli.py]: Syslog goes through UDP connect. There are any errors or warnings for no Syslog server.

File: helm/qontract-reconcile/templates/template.yaml (hunk 2 above)
Comment (BumbleFeng) [reply helm/qontract-reconcile/templates/template.yaml]: I will change it

File: reconcile/cli.py (hunk 1 above)
Comment (tparikh) [reply reconcile/cli.py]: Do you mean to say there will be no errors?

Comment (BumbleFeng) [reply reconcile/cli.py]: Yes.

File: reconcile/cli.py [hunk 4]
```diff
@@ -219,6 +220,14 @@ def init_log_level(log_level):
     logging.basicConfig(format=format,
                         datefmt='%Y-%m-%d %H:%M:%S',
                         level=level)
+    if 'syslog' in config.get_config():
```
Comment (BumbleFeng) [inline reconcile/cli.py]: Check if Syslog server is configured

File: helm/qontract-reconcile/templates/template.yaml [hunk 5]
```diff
@@ -22,8 +22,8 @@ objects:
           app: qontract-reconcile
       spec:
         {{- if $integration.logs }}
-        initContainers:
         {{- if $integration.logs.slack }}
+        initContainers:
```
Comment (maorfr) [inline helm/qontract-reconcile/templates/template.yaml]: good catch!

File: reconcile/cli.py (hunk 4 above)
Comment (maorfr) [reply reconcile/cli.py]: all integrations get the same config file. that means that all integrations will attempt to send logs to syslog.
unless i'm missing something?

File: reconcile/cli.py [hunk 6]
```diff
@@ -219,6 +220,14 @@ def init_log_level(log_level):
...
+        syslog_server = config.get_config()['syslog']['server']
+        syslog_port = int(config.get_config()['syslog']['port'])
```
Comment (maorfr) [inline reconcile/cli.py]: `config.get_config()` should be called once

File: reconcile/cli.py [hunk 7]
```diff
@@ -219,6 +220,14 @@ def init_log_level(log_level):
...
+        syslog = SysLogHandler(address=(syslog_server, syslog_port))
+        formatter = logging.Formatter(format, datefmt='%Y-%m-%d %H:%M:%S')
```
Comment (maorfr) [inline reconcile/cli.py]: can we extract the date format to a variable and re-use it?

Comment (maorfr) [reply reconcile/cli.py]: also, does this mean we are shipping logs from all levels?

File: reconcile/cli.py [hunk 8]
```diff
@@ -229,8 +238,8 @@ def init_log_level(log_level):
 def integration(ctx, configfile, dry_run, log_level):
     ctx.ensure_object(dict)
 
-    init_log_level(log_level)
     config.init_from_toml(configfile)
+    init_log_level(log_level)
```
Comment (maorfr) [inline reconcile/cli.py]: i think we should start thinking about sretoolbox.logging
out of scope of this PR of course.
@apahim, wdyt?

also feel free to take a look at this PR!

**Comment by apahim** [comment]
Hi @BumbleFeng, awesome work here. I like the approach you're taking, using Python logging, but it diverges from what we already have in place. My suggestion is to keep it consistent by using the same approach for both fluentd and cloudwatch.
Question: can we just extend the fluentd configuration to also send the log messages to cloudwatch, without the need for a syslog server side-car? Or even to have a side-car, but a fluentd one that looks almost the same, just with the cloudwatch config instead of a slack config?
I'm thinking of https://github.com/fluent-plugins-nursery/fluent-plugin-cloudwatch-logs

File: reconcile/cli.py (hunk 7 above)
Comment (tparikh) [reply reconcile/cli.py]: @BumbleFeng @maorfr Raised a good point. We should make log level for CloudWatch configurable. Debug logs may contain messages that should not be stored in external systems. We should make sure only log messages at `INFO` level or above are shipped to CloudWatch.

**Comment by tparikh** [comment]
If we were to use Fluentd, we should look into adding k8s metadata to log messages using filter https://github.com/fabric8io/fluent-plugin-kubernetes_metadata_filter

File: reconcile/cli.py (hunk 7 above)
Comment (BumbleFeng) [reply reconcile/cli.py]: `syslog.setLevel(log_level)` should work

File: reconcile/cli.py (hunk 4 above)
Comment (BumbleFeng) [reply reconcile/cli.py]: Yes. Syslog goes through UDP connect. So there are no errors or warnings for no Syslog server. Maybe we should add a new config flag to make Syslog optional.

**Comment by BumbleFeng** [comment]
Good idea. I will dive in Fluentd config.

**Comment by BumbleFeng** [comment]
Check fluentd solution here
https://github.com/app-sre/qontract-reconcile/pull/620

**Comment by maorfr** [comment]
Why did you choose to use `@copy` over a cloud watch fluentd plugin?

**Comment by BumbleFeng** [comment]
In case we need to ship the same log to both slack and cloudwatch. 

> Why did you choose to use `@copy` over a cloud watch fluentd plugin?

**Comment by maorfr** [comment]
Have you seen this?
https://github.com/fluent-plugins-nursery/fluent-plugin-cloudwatch-logs
As suggested by @apahim

**Comment by tparikh** [comment]
@maorfr he is using the right plugin. See https://github.com/app-sre/qontract-reconcile/pull/620/files#diff-56aa24bda0547277ddfbaffb78f5bf99R96

Copy sends logs to both slack and cloudwatch.
I would propose moving discussion to https://github.com/app-sre/qontract-reconcile/pull/620

**Comment by maorfr** [comment]
Sorry, misread. Ignore me.

**Pull request closed without merging**